File: bl/runners/http.py
"""
bl/runners/http.py — Simple HTTP request runner (C-07).

Executes one or more HTTP requests specified in the question's Test field
and returns a verdict envelope based on status code, response body, and latency.

Handles questions with mode: http

Test field syntax (parsed from the question):
    GET {url}                         — basic reachability
    POST {url} {json_body}            — POST with JSON body
    expect_status: {code}             — expected HTTP status (default 200)
    expect_body: {substring}          — response body must contain this
    latency_threshold_ms: {N}         — FAILURE if response > N ms
    auth: bearer                      — include AUTH_HEADERS from config
"""

import json
import re
import time

import httpx

from bl.config import auth_headers, cfg

# ---------------------------------------------------------------------------
# Spec parser
# ---------------------------------------------------------------------------

_URL_RE = re.compile(r"https?://\S+")
_METHOD_RE = re.compile(r"^\s*(GET|POST|PUT|DELETE|PATCH)\s+", re.IGNORECASE)


def _parse_http_spec(test_field: str) -> dict:
    """Parse the Test field of a question into an HTTP request spec.

    Returns a dict with keys:
        method, url, body, expect_status, expect_body, latency_ms, use_auth
    """
    spec = {
        "method": "GET",
        "url": None,
        "body": None,
        "expect_status": 200,
        "expect_body": None,
        "latency_ms": 2000,
        "use_auth": False,
    }

    lines = test_field.splitlines()

    for line in lines:
        stripped = line.strip()
        if not stripped:
            continue

        low = stripped.lower()

        # Directive lines
        if low.startswith("expect_status:"):
            try:
                spec["expect_status"] = int(stripped.split(":", 1)[1].strip())
            except ValueError:
                pass
            continue

        if low.startswith("expect_body:"):
            spec["expect_body"] = stripped.split(":", 1)[1].strip()
            continue

        if low.startswith("latency_threshold_ms:"):
            try:
                spec["latency_ms"] = int(stripped.split(":", 1)[1].strip())
            except ValueError:
                pass
            continue

        if low.startswith("auth:") and "bearer" in low:
            spec["use_auth"] = True
            continue

        # Method + URL line
        url_match = _URL_RE.search(stripped)
        if url_match and spec["url"] is None:
            spec["url"] = url_match.group(0).rstrip("/.,;")
            method_match = _METHOD_RE.match(stripped)
            if method_match:
                spec["method"] = method_match.group(1).upper()

            # Body: remainder of line after URL
            remainder = stripped[url_match.end() :].strip()
            if remainder:
                spec["body"] = remainder

    # Fallback URL
    if spec["url"] is None:
        spec["url"] = cfg.base_url.rstrip("/") + "/health"

    return spec


# ---------------------------------------------------------------------------
# Main runner
# ---------------------------------------------------------------------------


def run_http(question: dict) -> dict:
    """Execute HTTP request(s) from Test field and return verdict envelope."""
    test_field = question.get("test", "") or question.get("Test", "")
    spec = _parse_http_spec(test_field)

    method = spec["method"]
    url = spec["url"]
    expect_status = spec["expect_status"]
    latency_threshold = spec["latency_ms"]
    expect_body = spec["expect_body"]

    headers = {"Content-Type": "application/json"}
    if spec["use_auth"]:
        headers.update(auth_headers())

    body = None
    if spec["body"]:
        try:
            body = json.loads(spec["body"])
        except json.JSONDecodeError:
            body = spec["body"]

    # Execute request
    status_code = 0
    response_body = ""
    elapsed_ms = 0

    try:
        start = time.monotonic()
        with httpx.Client(timeout=cfg.request_timeout) as client:
            if method == "GET":
                resp = client.get(url, headers=headers)
            elif method == "POST":
                resp = client.post(
                    url,
                    headers=headers,
                    json=body if isinstance(body, dict) else None,
                    content=body.encode() if isinstance(body, str) else None,
                )
            elif method == "PUT":
                resp = client.put(
                    url, headers=headers, json=body if isinstance(body, dict) else None
                )
            elif method == "DELETE":
                resp = client.delete(url, headers=headers)
            else:
                resp = client.get(url, headers=headers)

        elapsed_ms = round((time.monotonic() - start) * 1000, 1)
        status_code = resp.status_code
        response_body = resp.text

    except httpx.ConnectError as exc:
        return {
            "verdict": "INCONCLUSIVE",
            "summary": f"HTTP {method} {url} — connection error: {exc}",
            "data": {"method": method, "url": url, "error": str(exc)},
            "details": f"Connection error reaching {url}: {exc}",
        }
    except httpx.TimeoutException as exc:
        return {
            "verdict": "INCONCLUSIVE",
            "summary": f"HTTP {method} {url} — timeout after {cfg.request_timeout}s",
            "data": {"method": method, "url": url, "error": "timeout"},
            "details": f"Request timed out after {cfg.request_timeout}s: {exc}",
        }
    except Exception as exc:
        return {
            "verdict": "INCONCLUSIVE",
            "summary": f"HTTP {method} {url} — unexpected error: {exc}",
            "data": {"method": method, "url": url, "error": str(exc)},
            "details": f"Unexpected error: {exc}",
        }

    # Determine verdict
    verdict = "HEALTHY"
    failure_reasons = []

    if status_code != expect_status:
        verdict = "FAILURE"
        failure_reasons.append(f"status {status_code} != expected {expect_status}")

    if elapsed_ms > latency_threshold:
        verdict = "FAILURE"
        failure_reasons.append(
            f"latency {elapsed_ms}ms > threshold {latency_threshold}ms"
        )
    elif verdict == "HEALTHY" and elapsed_ms > latency_threshold * 0.7:
        verdict = "WARNING"
        failure_reasons.append(
            f"latency {elapsed_ms}ms approaching threshold {latency_threshold}ms (>70%)"
        )

    if expect_body is not None and expect_body not in response_body:
        verdict = "FAILURE"
        failure_reasons.append(f"expected body substring '{expect_body}' not found")

    summary = f"HTTP {method} {url} → {status_code} in {elapsed_ms}ms"
    if failure_reasons:
        summary += " | " + "; ".join(failure_reasons)

    return {
        "verdict": verdict,
        "summary": summary,
        "data": {
            "method": method,
            "url": url,
            "status_code": status_code,
            "elapsed_ms": elapsed_ms,
            "latency_threshold_ms": latency_threshold,
            "body_preview": response_body[:200],
        },
        "details": (
            f"Status: {status_code} (expected {expect_status})\n"
            f"Latency: {elapsed_ms}ms (threshold: {latency_threshold}ms)\n"
            f"Body: {response_body[:300]}"
        ),
    }

File: masonry/src/engine/runners/http.js
'use strict';
// engine/runners/http.js — Simple HTTP request runner.
//
// Port of bl/runners/http.py to Node.js.

const { cfg } = require('../config');

const _URL_RE = /https?:\/\/\S+/;
const _METHOD_RE = /^\s*(GET|POST|PUT|DELETE|PATCH)\s+/i;

function _parseHttpSpec(testField) {
  const spec = {
    method: 'GET',
    url: null,
    body: null,
    expectStatus: 200,
    expectBody: null,
    latencyMs: 2000,
    useAuth: false,
  };

  for (const line of testField.split('\n')) {
    const stripped = line.trim();
    if (!stripped) continue;
    const low = stripped.toLowerCase();

    if (low.startsWith('expect_status:')) {
      const val = parseInt(stripped.slice(stripped.indexOf(':') + 1).trim(), 10);
      if (!isNaN(val)) spec.expectStatus = val;
      continue;
    }
    if (low.startsWith('expect_body:')) {
      spec.expectBody = stripped.slice(stripped.indexOf(':') + 1).trim();
      continue;
    }
    if (low.startsWith('latency_threshold_ms:')) {
      const val = parseInt(stripped.slice(stripped.indexOf(':') + 1).trim(), 10);
      if (!isNaN(val)) spec.latencyMs = val;
      continue;
    }
    if (low.startsWith('auth:') && low.includes('bearer')) {
      spec.useAuth = true;
      continue;
    }

    const urlMatch = stripped.match(_URL_RE);
    if (urlMatch && spec.url === null) {
      spec.url = urlMatch[0].replace(/[/.,;]+$/, '');
      const methodMatch = stripped.match(_METHOD_RE);
      if (methodMatch) {
        spec.method = methodMatch[1].toUpperCase();
      }
      const remainder = stripped.slice(urlMatch.index + urlMatch[0].length).trim();
      if (remainder) {
        spec.body = remainder;
      }
    }
  }

  if (spec.url === null) {
    spec.url = (cfg.baseUrl || 'http://localhost:8200').replace(/\/+$/, '') + '/health';
  }

  return spec;
}

function _buildHttpVerdict({ method, url, statusCode, elapsedMs, responseBody, expectStatus, latencyThreshold, expectBody }) {
  let verdict = 'HEALTHY';
  const failureReasons = [];

  if (statusCode !== expectStatus) {
    verdict = 'FAILURE';
    failureReasons.push(`status ${statusCode} != expected ${expectStatus}`);
  }

  if (elapsedMs > latencyThreshold) {
    verdict = 'FAILURE';
    failureReasons.push(`latency ${elapsedMs}ms > threshold ${latencyThreshold}ms`);
  } else if (verdict === 'HEALTHY' && elapsedMs > latencyThreshold * 0.7) {
    verdict = 'WARNING';
    failureReasons.push(`latency ${elapsedMs}ms approaching threshold ${latencyThreshold}ms (>70%)`);
  }

  if (expectBody !== null && !responseBody.includes(expectBody)) {
    verdict = 'FAILURE';
    failureReasons.push(`expected body substring '${expectBody}' not found`);
  }

  let summary = `HTTP ${method} ${url} → ${statusCode} in ${elapsedMs}ms`;
  if (failureReasons.length) {
    summary += ' | ' + failureReasons.join('; ');
  }

  return {
    verdict,
    summary,
    data: {
      method,
      url,
      status_code: statusCode,
      elapsed_ms: elapsedMs,
      latency_threshold_ms: latencyThreshold,
      body_preview: responseBody.slice(0, 200),
    },
    details: `Status: ${statusCode} (expected ${expectStatus})\nLatency: ${elapsedMs}ms (threshold: ${latencyThreshold}ms)\nBody: ${responseBody.slice(0, 300)}`,
  };
}

async function runHttp(question) {
  const testField = question.test || question.Test || '';
  const spec = _parseHttpSpec(testField);

  const { method, url, expectStatus, latencyMs: latencyThreshold, expectBody } = spec;

  const headers = { 'Content-Type': 'application/json' };
  let body = null;
  if (spec.body) {
    try {
      body = JSON.parse(spec.body);
    } catch {
      body = spec.body;
    }
  }

  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), (cfg.requestTimeout || 30) * 1000);

    const fetchOpts = { method, headers, signal: controller.signal };
    if (body && method !== 'GET') {
      fetchOpts.body = typeof body === 'string' ? body : JSON.stringify(body);
    }

    const start = Date.now();
    const resp = await fetch(url, fetchOpts);
    const elapsedMs = Date.now() - start;
    clearTimeout(timeoutId);

    const responseBody = await resp.text();

    return _buildHttpVerdict({
      method, url,
      statusCode: resp.status,
      elapsedMs,
      responseBody,
      expectStatus,
      latencyThreshold,
      expectBody,
    });
  } catch (err) {
    if (err.name === 'AbortError') {
      return {
        verdict: 'INCONCLUSIVE',
        summary: `HTTP ${method} ${url} — timeout after ${cfg.requestTimeout || 30}s`,
        data: { method, url, error: 'timeout' },
        details: `Request timed out after ${cfg.requestTimeout || 30}s`,
      };
    }
    return {
      verdict: 'INCONCLUSIVE',
      summary: `HTTP ${method} ${url} — ${err.message}`,
      data: { method, url, error: err.message },
      details: `Error: ${err.message}`,
    };
  }
}

module.exports = {
  runHttp,
  _parseHttpSpec,
  _buildHttpVerdict,
};
